File: LeetCode/Medium/739. Daily Temperatures.py
class Solution:
    
    def dailyTemperatures( self, temperatures: List[int] ) -> List[int]:
        
        ans = [ 0 ] * len( temperatures )
        a = [ 0 ]
        
        for i in range( 1, len( temperatures ) ):
            while len( a ) > 0 and temperatures[ a[-1] ] < temperatures[i]:
                k = a.pop()
                ans[k] = i - k
            
            a.append( i )
                    
        return ans

File: LeetCode/Medium/739. Daily Temperatures.java
import java.util.AbstractMap.SimpleImmutableEntry;

class Solution
{
    
    public int[] dailyTemperatures
    ( int[] temperatures )
    {
        int n = temperatures.length;
        int[] ans = new int[n];
        Stack< SimpleImmutableEntry< Integer, Integer > > stack =
            new Stack<>();
            
        for ( int i = 0; i < n; i++ )
        {
            while ( !stack.isEmpty()
                 && stack.peek().getValue() < temperatures[i] )
            {
                int j = stack.pop().getKey();
                ans[j] = i - j;
            }
            
            stack.push( new SimpleImmutableEntry<>( i, temperatures[i] ) );
        }
        
        return ans;
    }
    
}
